components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/persistence/JDBCPersistenceManager.java

```diff
@@ -99,13 +105,28 @@ private void initDataSource() {
             }
 
             OMElement dataSourceNameElem = dataSourceElem.getFirstChildWithName(
-                    new QName(IdentityCoreConstants.IDENTITY_DEFAULT_NAMESPACE, "Name"));
+                    new QName(IdentityCoreConstants.IDENTITY_DEFAULT_NAMESPACE, NAME));
 
             if (dataSourceNameElem != null) {
                 String dataSourceName = dataSourceNameElem.getText();
                 Context ctx = new InitialContext();
                 dataSource = (DataSource) ctx.lookup(dataSourceName);
             }
+            OMElement sessionPersistElem = persistenceManagerConfigElem.getFirstChildWithName(
+                    new QName(IdentityCoreConstants.IDENTITY_DEFAULT_NAMESPACE, SESSION_DATA_PERSIST));
+            OMElement sessionDataSourceElem = sessionPersistElem.getFirstChildWithName(
+                    new QName(IdentityCoreConstants.IDENTITY_DEFAULT_NAMESPACE, DATA_SOURCE));
+            if (sessionDataSourceElem != null) {
+                OMElement sessionDataSourceNameElem = sessionDataSourceElem.getFirstChildWithName(
+                        new QName(IdentityCoreConstants.IDENTITY_DEFAULT_NAMESPACE, NAME));
+                if (sessionDataSourceNameElem != null) {
+                    String dataSourceName = sessionDataSourceNameElem.getText();
+                    Context ctx = new InitialContext();
+                    sessionDataSource = (DataSource) ctx.lookup(dataSourceName);
+                }
+            } else {
+                sessionDataSource = dataSource;
```

Comment:
Better have a debug log 
if(log.isDebugEnabled) {
log.debug("Using default identity datasource since the different session data source is not configured")
}

Reply:
done